### DISCUSSION TITLE
Is metaprogramming feature set of Nelua more powerful than that of Common Lisp or Nim?

### DISCUSSION THREAD

**Discussion by przemyslaw0** [edubart/nelua-lang#218] [Q&A] [answered]
I have been seeking for a extensible programming language that would let me extend as I wish.

Some say that Common Lisp has the most featurefull set of metaprogramming facilities, but has Nelua a equal or more complete metaprogramming feature set? Could I do all the metaprogramming things that Common Lisp let to do?

Also, I know that metaprogramming of Nim is powerful. It has support for [Generalized raw string literals](https://nim-lang.org/docs/manual.html#lexical-analysis-generalized-raw-string-literals), something similar of that of Julia language, called ["string macros"](https://stackoverflow.com/questions/26600143/why-do-string-macros-in-julia-use). Has Nelua something similar to generalized raw strings literals or "string macros"?

**Comment by asumbek** [comment]
> has Nelua a equal or more complete metaprogramming feature set? Could I do all the metaprogramming things that Common Lisp let to do?

Seems not, but you can modify the AST using preprocessor.

> Has Nelua something similar to generalized raw strings literals or "string macros"?

Seems not, did you mean the old Lua syntax that allows you to omit parentheses? (e.g `print"Hello World!"`)

**Reply by przemyslaw0** [reply]
>Seems not, did you mean the old Lua syntax that allows you to omit parentheses? (e.g print"Hello World!")

Something like, it's that feature called ["reader macros"](https://www.offerzen.com/blog/how-reader-macros-enhance-lisps-expressive-power) in Common Lisp, that let the programmer invent his own syntax.

**Reply by stefanos82** [reply, marked as answer]
Go in `examples` and see these files:

![image](https://github.com/edubart/nelua-lang/assets/1777466/7e131252-a1dd-4be8-aea7-3a92cc27f062)

Prefer to see `dsl2` example first and you should find all the answers you are looking for.

**Reply by przemyslaw0** [reply]
@stefanos82 

Thank you very much Man, I have seen the "grammar" extension, it allows a thing like reader macros in Common Lisp.

I will choose Nelua PL as one of my programming languages and I'm expecting to write parsers, libraries and other stuff in Nelua in the next 1~2 years.

**Reply by stefanos82** [reply]
Don't forget to share the fun with us :smiley: it really feels motivating witnessing other developers using Nelua for any type of project.

**Comment by przemyslaw0** [comment]
@stefanos82

I have a lot of things to do before starting with Nelua programming, I think I will need to learn Lua before.

I have a question as you're one of the maintainers of this project.

Is it possible to use non-C features like classes in the Nelua file when importing a C++/Objective-C file or header? I know that Nelua has no OOP support, but is it possible to write a macro to use the classes of a C++/Objective-C file in Nelua code?

I found a lot of interesting things about that in Lua, but I don't know if it's possible to do the same with Nelua:

http://lua-users.org/wiki/BindingCodeToLua
https://olua.sourceforge.net/

==============

I will focus on making interoperability of Nelua with C-trio (C/C++/Objective-C) when I master Lua/Nelua.

**Comment by stefanos82** [comment]
> I have a lot of things to do before starting with Nelua programming, I think I will need to learn Lua before.

You can learn [Lua in 15 minutes](https://tylerneylon.com/a/learn-lua/), don't worry about this.

> I have a question as you're one of the maintainers of this project.

I'm not a maintainer, I'm just a user like yourself who happens to be extremely curious around programming languages.

> Is it possible to use non-C features like classes in the Nelua file when importing a C++/Objective-C file or header?

This is something only @edubart can answer, I have no idea how to do this; only with C I know.

> I know that Nelua has no OOP support, but is it possible to write a macro to use the classes of a C++/Objective-C file in Nelua code?

Nelua has support for OOP but it uses `record`s instead of `class`es. I'm sure there is a way.

> I found a lot of interesting things about that in Lua, but I don't know if it's possible to do the same with Nelua:
> 
> http://lua-users.org/wiki/BindingCodeToLua https://olua.sourceforge.net/

Again, you can do pretty much anything you want with Nelua, as it uses Lua as its preprocessor behind the scenes; you can extend it in any way you like.

If you read the entire https://nelua.io/overview/ along with https://github.com/edubart/nelua-lang/blob/master/examples/overview.nelua plus all the https://github.com/edubart/nelua-lang/blob/master/examples/ , then you have nothing to worry about.

**Reply by przemyslaw0** [reply]
Thanks for the time answering me, @stefanos82.

@edubart Could you please answer my question? I tried to enter in contact with you but got no response.
